### ISSUE TITLE
Bug in initialisation of species

### ISSUE THREAD

**Issue by hivert** [sagemath/sage#12649]
<div id="comment:0"></div>

The following specification is valid:

```
sage: Set0  = species.SetSpecies()
sage: Set1  = species.SetSpecies(min=1)
sage: Set2  = species.SetSpecies(min=2)
sage: Four = species.CombinatorialSpecies()
sage: Four.define(Set2(Four) + Set1*Set0(Four))
```
However, the following code breaks

```
sage: Four.generating_series()[1]
[...]
ValueError: generator already executing
```
Strangely enough, adding some seemingly unrelated instruction work
around it::

```
sage: Set0  = species.SetSpecies()
sage: Set1  = species.SetSpecies(min=1)
sage: Set2  = species.SetSpecies(min=2)

sage: Set2(Set1).generating_series()[1]
0

sage: Four = species.CombinatorialSpecies()
sage: Four.define(Set2(Four) + Set1*Set0(Four))
sage: Four.generating_series()[1]
1
sage: Four.generating_series()[4]
52/3
sage: Four.generating_series()
x + 2*x^2 + 16/3*x^3 + 52/3*x^4 + O(x^5)
```

Florent

CC:  @sagetrac-sage-combinat @mantepse

Component: **combinatorics**

Keywords: **Species, LazyPowerSeries**

_Issue created by migration from https://trac.sagemath.org/ticket/12649_

**Comment by fchapoton** [comment]
<div id="comment:7" align="right">comment:7</div>

will this be fixed by the recent refactoring ?

maybe check also other tickets with "species" keyword ?

**Comment by mantepse** [comment]
<div id="comment:8" align="right">comment:8</div>

Yes, this already works.  However, you have to be explicit about the valuation of `Four` now, because it cannot determine that it doesn't have valuation 0.

(I admit I don't see right now whether it should be able to determine that the valuation is not 0 - do you?)

```
sage: Set0  = species.SetSpecies()
sage: Set1  = species.SetSpecies(min=1)
sage: Set2  = species.SetSpecies(min=2)
sage: Four = species.CombinatorialSpecies(min=1)
sage: Four.define(Set2(Four) + Set1*Set0(Four))

sage: Four.generating_series()
z + 2*z^2 + 16/3*z^3 + 52/3*z^4 + 944/15*z^5 + 11008/45*z^6 + 313664/315*z^7 + O(z^8)

sage: Four.isotype_generating_series()
z + 3*z^2 + 10*z^3 + 40*z^4 + 170*z^5 + 785*z^6 + 3770*z^7 + O(z^8)
sage: Four.cycle_index_series()
p[1] + (2*p[1,1]+p[2]) + (16/3*p[1,1,1]+4*p[2,1]+2/3*p[3]) + (52/3*p[1,1,1,1]+16*p[2,1,1]+3*p[2,2]+8/3*p[3,1]+p[4]) + (944/15*p[1,1,1,1,1]+208/3*p[2,1,1,1]+20*p[2,2,1]+32/3*p[3,1,1]+8/3*p[3,2]+4*p[4,1]+2/5*p[5]) + (11008/45*p[1,1,1,1,1,1]+944/3*p[2,1,1,1,1]+120*p[2,2,1,1]+12*p[2,2,2]+416/9*p[3,1,1,1]+64/3*p[3,2,1]+14/9*p[3,3]+16*p[4,1,1]+6*p[4,2]+8/5*p[5,1]+p[6]) + (313664/315*p[1,1,1,1,1,1,1]+22016/15*p[2,1,1,1,1,1]+2096/3*p[2,2,1,1,1]+112*p[2,2,2,1]+1888/9*p[3,1,1,1,1]+416/3*p[3,2,1,1]+40/3*p[3,2,2]+88/9*p[3,3,1]+208/3*p[4,1,1,1]+40*p[4,2,1]+8/3*p[4,3]+32/5*p[5,1,1]+8/5*p[5,2]+4*p[6,1]+2/7*p[7]) + O^8
```

**Comment by mantepse** [comment]
<div id="comment:9" align="right">comment:9</div>

Yes, I am eager to close all the other tickets.  However, I want #34552 to be in the develop branch first.

**Comment by mantepse** [comment]
<div id="comment:10" align="right">comment:10</div>

Hm, I think it should know that the valuation is greater than 0.  The valuation of `Set1 * Set0(Four)` is greater than zero, because the valuation of `Set1` is.

`Set2` is not a polynomial (because it is not 'exact'), so `Set2.__call__` is only defined if its argument has valuation at least 1.  Thus we might as well assume that its valuation is at least 1.

I'm not sure yet.

**Comment by mantepse** [comment]
Changed keywords from **Species** to **Species, LazyPowerSeries**

**Comment by mantepse** [comment]
<div id="comment:12" align="right">comment:12</div>

We actually thought about this situation (I forgot), and introduced a `check` parameter for `__call__`.  Of course, it is not passed by the species code.

I am still not sure what the best solution might be.

**Comment by mantepse** [comment]
Fixed by #36284.
